### PR TITLE
Added PRISMA_BINARIES_MIRROR example

### LINKED ISSUE #447: Document PRISMA_BINARIES_MIRROR
See https://github.com/prisma/prisma/issues/1333#issuecomment-607868654

Implemented a new ENV var `PRISMA_BINARIES_MIRROR` that also changes the download location that is used to get the binaries.

### LINKED ISSUE #1425: Document how to get relevant binaries that mirror url
We are missing documentation (and concept!) how to get all the relevant binaries at that mirror URL though.

See [this comment](https://github.com/prisma/docs/pull/1422#discussion_r595103665)

### PR DESCRIPTION
**Pull request by molebox**
This PR adds a bit more weight to how to use this environment variable with an example in the concepts section which is linked to the API references.

Fixes: #447 
Fixes: #1425

### CODE REVIEW

File: content/200-concepts/050-overview/500-under-the-hood.mdx [hunk 1]
```diff
@@ -105,24 +116,24 @@ The output shows that the query engine path comes from the `PRISMA_QUERY_ENGINE_
...
    <tab>
 
-   ```sh highlight=2;normal
-   Current platform     : darwin
-   Query Engine         : query-engine d6ff7119649922b84e413b3b69660e2f49e2ddf3 (at /custom/my-query-engine-unix)
-   Migration Engine     : migration-engine-cli d6ff7119649922b84e413b3b69660e2f49e2ddf3 (at /myproject/node_modules/@prisma/engines/migration-engine-unix)
-   Introspection Engine : introspection-core d6ff7119649922b84e413b3b69660e2f49e2ddf3 (at /myproject/node_modules/@prisma/engines/introspection-engine-unix)
-   Format Binary        : prisma-fmt d6ff7119649922b84e413b3b69660e2f49e2ddf3 (at /myproject/node_modules/@prisma/engines/prisma-fmt-unix)
-   ```
+```sh highlight=2;normal
```
Comment (janpio) [inline content/200-concepts/050-overview/500-under-the-hood.mdx:108]: Is this change in indentation intentional?

File: content/200-concepts/050-overview/500-under-the-hood.mdx [hunk 2]
```diff
@@ -75,22 +77,31 @@ Run the following command to set the environment variable globally (in this exam
...
-   ```
+```terminal
+set PRISMA_QUERY_ENGINE_BINARY=c:\custom\my-query-engine-windows.exe
+```
 
    </tab>
 
 </TabbedContent>
 
+#### Navigating a corperate firewall with <inlinecode>PRISMA_BINARIES_MIRROR</inlinecode>
```
Comment (janpio) [inline content/200-concepts/050-overview/500-under-the-hood.mdx]: This is only one of many use cases. This is much more guide-y than the concept documentation usually is.

Fundamentally you just
> ## Define a download mirror with `PRISMA_BINARIES_MIRROR`

File: content/200-concepts/050-overview/500-under-the-hood.mdx [hunk 3]
```diff
@@ -75,22 +77,31 @@ Run the following command to set the environment variable globally (in this exam
...
+
+You might find that your company operates behind a corporate firewall and you are unable to access the Prisma binaries via the default url `https://binaries.prisma.sh`.
+In such situations the `PRISMA_BINARIES_MIRROR` environment variable can be used to bypass the corporate firewall by proxying the request to the Prisma domain and download the binaries.
+
+```terminal
+PRISMA_BINARIES_MIRROR=https://my-whitelisted-url
+```
```
Comment (Jolg42) [inline content/200-concepts/050-overview/500-under-the-hood.mdx]: This is not my understanding of this feature, even though it could work.

I think proxying can be achieved with https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#proxy-environment-variables

`PRISMA_BINARIES_MIRROR` main goal is to provide a way for people to do the binary hosting themselves. For example, if a company would want to build their own binaries for any reason (custom platform or they want to run a fork of the binaries or for security point of view they want to build it on their hardware). Then they can create for example an AWS S3 bucket and upload their binaries.

Comment (janpio) [reply content/200-concepts/050-overview/500-under-the-hood.mdx]: You _can_ create a "proxy" in front of our domain with your own domain to get around firewall rules and so on. But that is just one of multiple use cases of this variable and probably defeats the purpose of the security rule.

**Review by janpio — CHANGES_REQUESTED** [review]
see comments

File: content/200-concepts/050-overview/500-under-the-hood.mdx [hunk 4]
```diff
@@ -30,11 +30,13 @@ By default, all binaries are automatically downloaded into the `node_modules/@pr
 
 - Automated download of binaries is not possible.
 - You have created your own engine binary (for testing, or for an OS that is not officially supported).
+- You are behind a corperate firewall and need to use a whitelisted url to download the Prisma binaries.
```
Comment (janpio) [inline content/200-concepts/050-overview/500-under-the-hood.mdx]: This is not related to using custom engine binaries (headline)

File: content/200-concepts/050-overview/500-under-the-hood.mdx [hunk 5]
```diff
@@ -30,11 +30,13 @@ By default, all binaries are automatically downloaded into the `node_modules/@pr
...
 
 Use the following environment variables to specify custom locations for your binaries:
 
+- [`PRISMA_BINARIES_MIRROR`](../../reference/api-reference/environment-variables-reference/#prisma_binaries_mirror) <span class="api"></span>
```
Comment (janpio) [inline content/200-concepts/050-overview/500-under-the-hood.mdx]: This is only indirectly related to using custom engine binaries (headline)

File: content/400-reference/200-api-reference/300-environment-variables-reference.mdx [hunk 6]
```diff
@@ -97,38 +96,40 @@ By default all binaries are downloaded when you install Prisma CLI. There are ho
 
 ### <inlinecode>PRISMA_BINARIES_MIRROR</inlinecode>
 
-`PRISMA_BINARIES_MIRROR` can be used to specify a custom CDN endpoint (or any server) to download the engines binaries for the CLI/Client. The default value is `https://binaries.prisma.sh`.
+`PRISMA_BINARIES_MIRROR` can be used to specify a custom CDN (or server) endpoint to download the engines binaries for the CLI/Client. The default value is `https://binaries.prisma.sh`.
+
+See [`Navigating a corporate firewall`](../../concepts/overview/under-the-hood#navigating-a-corperate-firewall-with-prisma_binaries_mirror) for a conceptual overview of how to use this environment variable.
```
Comment (janpio) [inline content/400-reference/200-api-reference/300-environment-variables-reference.mdx]: "Navitationg a coporate firewall" would be an example for a problem to work around using that env var.

File: content/400-reference/200-api-reference/300-environment-variables-reference.mdx [hunk 7]
```diff
@@ -97,38 +96,40 @@ By default all binaries are downloaded when you install Prisma CLI. There are ho
...
 
 ### <inlinecode>PRISMA_MIGRATION_ENGINE_BINARY</inlinecode>
 
 `PRISMA_MIGRATION_ENGINE_BINARY` is used to set a custom location for your own migration engine binary.
 
 ```js
-PRISMA_MIGRATION_ENGINE_BINARY=custom/my-migration-engine-unix
+PRISMA_MIGRATION_ENGINE_BINARY = custom / my - migration - engine - unix
```
Comment (janpio) [inline content/400-reference/200-api-reference/300-environment-variables-reference.mdx]: this is broken

File: content/400-reference/200-api-reference/300-environment-variables-reference.mdx [hunk 8]
```diff
@@ -97,38 +96,40 @@ By default all binaries are downloaded when you install Prisma CLI. There are ho
...
+PRISMA_INTROSPECTION_ENGINE_BINARY = custom / my - introspection - engine - unix
 ```
 
 ### <inlinecode>PRISMA_FMT_BINARY</inlinecode>
 
 `PRISMA_FMT_BINARY` is used to set a custom location for your own format engine binary.
 
 ```js
-PRISMA_FMT_BINARY=custom/my-custom-format-engine-unix
+PRISMA_FMT_BINARY = custom / my - custom - format - engine - unix
```
Comment (janpio) [inline content/400-reference/200-api-reference/300-environment-variables-reference.mdx]: these edits are broken

File: content/200-concepts/050-overview/500-under-the-hood.mdx (hunk 1 above)
Comment (molebox) [reply content/200-concepts/050-overview/500-under-the-hood.mdx:108]: Not intentional. Must have been prettier formatting it

File: content/200-concepts/050-overview/500-under-the-hood.mdx (hunk 2 above)
Comment (molebox) [reply content/200-concepts/050-overview/500-under-the-hood.mdx]: Hmm. Do we want to give any examples in the ceoncepts section then? As im learning what this does, this is the use case that came up when we were speaking so thats the one i used. I feel like an explanation is needed as to what `Define a download mirror with PRISMA_BINARIES_MIRROR` means, hence the example.

If i was coming to this page for the first time i wouldnt have a clue what that meant.

File: content/400-reference/200-api-reference/300-environment-variables-reference.mdx (hunk 7 above)
Comment (molebox) [reply content/400-reference/200-api-reference/300-environment-variables-reference.mdx]: Argh, prettier

File: content/400-reference/200-api-reference/300-environment-variables-reference.mdx (hunk 8 above)
Comment (molebox) [reply content/400-reference/200-api-reference/300-environment-variables-reference.mdx]: I think my workspace prettier setup might be clashing with the docs repo one...

Comment (Jolg42) [reply content/400-reference/200-api-reference/300-environment-variables-reference.mdx]: Not sure if that helps but I see ```js and this is not valid js, maybe switch to another code tag (but it is valid bash)
```
PRISMA_FMT_BINARY=custom/my-custom-format-engine-unix
```

File: content/200-concepts/050-overview/500-under-the-hood.mdx [hunk 9]
```diff
@@ -75,22 +77,31 @@ Run the following command to set the environment variable globally (in this exam
...
 
 </TabbedContent>
 
+#### Navigating a corperate firewall with <inlinecode>PRISMA_BINARIES_MIRROR</inlinecode>
+
+You might find that your company operates behind a corporate firewall and you are unable to access the Prisma binaries via the default url `https://binaries.prisma.sh`.
+In such situations the `PRISMA_BINARIES_MIRROR` environment variable can be used to bypass the corporate firewall by proxying the request to the Prisma domain and download the binaries.
+
+```terminal
+PRISMA_BINARIES_MIRROR=https://my-whitelisted-url
```
Comment (Jolg42) [inline content/200-concepts/050-overview/500-under-the-hood.mdx]: We prefer using alternatives to blacklist / whitelist. (racially insensitive terms).
Maybe change to allow / allowed

Comment (Jolg42) [reply content/200-concepts/050-overview/500-under-the-hood.mdx]: Check everywhere in PR to replace all occurences

File: content/200-concepts/050-overview/500-under-the-hood.mdx (hunk 2 above)
Comment (janpio) [reply content/200-concepts/050-overview/500-under-the-hood.mdx]: Then the explanation is that you can set a different base URL for the downloading of the binaries, that is used instead of our default URL (binaries.prisma.sh). That is the concept as I understand it.

"Navigating a corporate firewall" then might be one of the use cases or problems that can be solved with it.

File: content/400-reference/200-api-reference/300-environment-variables-reference.mdx (hunk 8 above)
Comment (molebox) [reply content/400-reference/200-api-reference/300-environment-variables-reference.mdx]: Oh thats a relic, i didnt write that. But i didnt notice it either so will change. Good catch!

File: content/200-concepts/050-overview/500-under-the-hood.mdx [hunk 10]
```diff
@@ -105,25 +105,34 @@ The output shows that the query engine path comes from the `PRISMA_QUERY_ENGINE_
...
+```
 
    </tab>
 
 </TabbedContent>
+
+### Hosting binaries
+
+The [`PRISMA_BINARIES_MIRROR`](../../reference/api-reference/environment-variables-reference/#prisma_binaries_mirror) <span class="api"></span> environment variable allows you to host a binary via a private server, AWS bucket or other cloud storage.
+This can be useful if you have a custom OS which uses your own binaries.
```
Comment (janpio) [inline content/200-concepts/050-overview/500-under-the-hood.mdx]:
```suggestion
This can be useful if you have a custom OS which requires you to use your own custom built binaries.
```

File: content/400-reference/200-api-reference/300-environment-variables-reference.mdx [hunk 11]
```diff
@@ -97,37 +96,43 @@ By default all binaries are downloaded when you install Prisma CLI. There are ho
 
 ### <inlinecode>PRISMA_BINARIES_MIRROR</inlinecode>
 
-`PRISMA_BINARIES_MIRROR` can be used to specify a custom CDN endpoint (or any server) to download the engines binaries for the CLI/Client. The default value is `https://binaries.prisma.sh`.
+`PRISMA_BINARIES_MIRROR` can be used to specify a custom CDN (or server) endpoint to download the engines binaries for the CLI/Client. The default value is `https://binaries.prisma.sh`.
```
Comment (janpio) [inline content/400-reference/200-api-reference/300-environment-variables-reference.mdx]:
```suggestion
`PRISMA_BINARIES_MIRROR` can be used to specify a custom CDN (or server) endpoint to download the engines binaries for the CLI/Client. The default value is `https://binaries.prisma.sh`, where Prisma hosts the binaries.
```

**Review by janpio — APPROVED** [review]

File: content/200-concepts/050-overview/500-under-the-hood.mdx [hunk 12]
```diff
@@ -105,25 +105,34 @@ The output shows that the query engine path comes from the `PRISMA_QUERY_ENGINE_
...
+Migration Engine     : migration-engine-cli d6ff7119649922b84e413b3b69660e2f49e2ddf3 (at c:\myproject\node_modules\@prisma\engines\migration-engine-windows.exe)
+Introspection Engine : introspection-core d6ff7119649922b84e413b3b69660e2f49e2ddf3 (at c:\myproject\node_modules\@prisma\engines\introspection-engine-windows.exe)
+Format Binary        : prisma-fmt d6ff7119649922b84e413b3b69660e2f49e2ddf3 (at c:\myproject\node_modules\@prisma\engines\prisma-fmt-windows.exe)
+```
 
    </tab>
 
 </TabbedContent>
+
+### Hosting binaries
```
Comment (janpio) [inline content/200-concepts/050-overview/500-under-the-hood.mdx:131]: Headline does not feel great, but unclear what would be better.

Comment (molebox) [reply content/200-concepts/050-overview/500-under-the-hood.mdx:131]: I agree. Ill have a chat with @mhwelander and see if we can come up with something better, in the meantime i will merge this. Thanks for all your help @janpio 🙌🏻
